src/java/commissioner.i

```diff
@@ -0,0 +1,153 @@
+ *    CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+ *    SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+ *    INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+ *    CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ *    ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ *    POSSIBILITY OF SUCH DAMAGE.
+ */
+
+/**
+ * @file
+ *   This file defines the SWIG interface of the commissioner.
+ *
+ * See http://www.swig.org for more information about SWIG.
+ *
+ */
+
+%module(directors="1") commissionerModule
+
+%{
+#include <commissioner/defines.hpp>
+#include <commissioner/error.hpp>
+#include <commissioner/network_data.hpp>
+#include <commissioner/commissioner.hpp>
+%}
+
+%include <std_string.i>
+%include <stdint.i>
+%include <std_list.i>
+%include <std_shared_ptr.i>
+%include <std_vector.i>
+%include <stl.i>
+%include <typemaps.i>
+
+// Remove the 'm' prefix of all members.
+%rename("%(regex:/^(m)(.*)/\\2/)s") "";
+
+// Insert the code of loading native shared library into generated Java class.
+%pragma(java) jniclasscode=%{
+    static {
+        try {
+            System.loadLibrary("commissioner-java");
+        } catch (UnsatisfiedLinkError e) {
+            System.err.println("failed to load native commissioner library!\n" + e);
+            System.exit(1);
+        }
+    }
+%}
+
+%feature("director") ot::commissioner::CommissionerHandler;
+%feature("director") ot::commissioner::Logger;
+
+%template(ByteArray) std::vector<uint8_t>;
+%template(ChannelMask) std::vector<ot::commissioner::ChannelMaskEntry>;
+%template(StringVector) std::vector<std::string>;
+
+%apply uint8_t& OUTPUT { uint8_t& aStatus }
+
+%typemap(jstype) std::string& OUTPUT "String[]"
+%typemap(jtype)  std::string& OUTPUT "String[]"
+%typemap(jni)    std::string& OUTPUT "jobjectArray"
+%typemap(javain) std::string& OUTPUT "$javainput"
+%typemap(in)     std::string& OUTPUT (std::string temp) {
+    if (!$input) {
+        SWIG_JavaThrowException(jenv, SWIG_JavaNullPointerException, "array null");
+        return $null;
+    }
+    if (JCALL1(GetArrayLength, jenv, $input) == 0) {
+        SWIG_JavaThrowException(jenv, SWIG_JavaIndexOutOfBoundsException, "Array must contain at least 1 element");
+    }
+    $1 = &temp;
+}
+%typemap(argout) std::string& OUTPUT {
+  jstring jvalue = JCALL1(NewStringUTF, jenv, temp$argnum.c_str()); 
+  JCALL3(SetObjectArrayElement, jenv, $input, 0, jvalue);
+}
+
+%apply std::string& OUTPUT { std::string &aMeshLocalAddr }
+%apply std::string& OUTPUT { std::string &aExistingCommissionerId }
+
+%shared_ptr(ot::commissioner::Logger)
+%shared_ptr(ot::commissioner::Commissioner)
+
+namespace ot {
+namespace commissioner {
+    // Remove async commissioner APIs.
+    %ignore Commissioner::Connect(ErrorHandler aHandler, const std::string &aAddr, uint16_t aPort);
```

Comment:
can we also list sync APIs so that we can tell which interfaces are provided just by this file.

Reply:
There is technology difficulty that SWIG doesn't support `std::function` which is callback of the async APIs. One possible workaround is to move those callbacks to `CommissionerHandler`. But, since it is not recommend to call async functions between Java runtime native C/C++, async APIs are not necessary for java.